File: packages/wazeko-auth/src/pairing.ts
import { PairingCodeEvent } from "../../wazeko-types/src/index.js";

export class PairingCodeManager {
  private currentCode: PairingCodeEvent | null = null;

  generateCode(expiresInSeconds: number = 120): PairingCodeEvent {
    const charset = "23456789ABCDEFGHJKLMNPQRSTUVWXYZ";
    let p1 = "";
    let p2 = "";

    for (let i = 0; i < 4; i++) {
      p1 += charset[Math.floor(Math.random() * charset.length)];
      p2 += charset[Math.floor(Math.random() * charset.length)];
    }

    const code = `${p1}-${p2}`;
    const event: PairingCodeEvent = {
      code,
      expiresInSeconds,
    };
    this.currentCode = event;
    return event;
  }

  printCode(): void {
    if (this.currentCode) {
      console.log("\n==============================");
      console.log(`  WHATSAPP PAIRING CODE: ${this.currentCode.code}`);
      console.log(`  Expires in: ${this.currentCode.expiresInSeconds}s`);
      console.log("==============================\n");
    }
  }

  get current(): PairingCodeEvent | null {
    return this.currentCode;
  }
}

File: crates/wazeko-auth/src/pairing.rs
use rand::Rng;
use tracing::info;
use wazeko_types::events::PairingCodeEvent;

pub struct PairingCodeManager {
    current_code: Option<PairingCodeEvent>,
}

impl Default for PairingCodeManager {
    fn default() -> Self {
        Self::new()
    }
}

impl PairingCodeManager {
    pub fn new() -> Self {
        Self { current_code: None }
    }

    pub fn generate_code(&mut self, expires_in_seconds: u64) -> PairingCodeEvent {
        let mut rng = rand::thread_rng();
        let charset = b"23456789ABCDEFGHJKLMNPQRSTUVWXYZ"; // WhatsApp pairing base32-like alphabet
        let p1: String = (0..4)
            .map(|_| {
                let idx = rng.gen_range(0..charset.len());
                charset[idx] as char
            })
            .collect();
        let p2: String = (0..4)
            .map(|_| {
                let idx = rng.gen_range(0..charset.len());
                charset[idx] as char
            })
            .collect();

        let formatted = format!("{p1}-{p2}");
        let event = PairingCodeEvent {
            code: formatted,
            expires_in_seconds,
        };
        self.current_code = Some(event.clone());
        event
    }

    pub fn print_code(&self) {
        if let Some(event) = &self.current_code {
            info!(target: "wazeko::auth::pairing", "Pairing Code: {}", event.code);
            println!("\n==============================");
            println!("  WHATSAPP PAIRING CODE: {}", event.code);
            println!("  Expires in: {}s", event.expires_in_seconds);
            println!("==============================\n");
        }
    }

    pub fn current(&self) -> Option<&PairingCodeEvent> {
        self.current_code.as_ref()
    }
}
